plugin/pkg/policy/expression.go

```diff
@@ -0,0 +1,118 @@
+package policy
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"github.com/coredns/coredns/request"
+
+	"github.com/coredns/coredns/plugin/metadata"
+
+	expr "github.com/Knetic/govaluate"
+)
+
+type policyRuleExpression struct {
+	kind        byte
+	kindIfError byte
+	expression  *expr.EvaluableExpression
+}
+
+// ExpressionEngine implement interface Engine for Firewall plugin
+// it evaluate the rues using an the lib Knetic/govaluate
+type ExpressionEngine struct {
+	kindIfErrorEvaluation byte
+	dataFromReq           *RequestExtractorMapping
+}
+
+type dataAsParam struct {
+	ctx         context.Context
+	dataFromReq *RequestDataExtractor
+}
+
+// NewExpressionEngine create a new Engine with default configuration
+func NewExpressionEngine() *ExpressionEngine {
+	return &ExpressionEngine{TypeRefuse, NewRequestExtractorMapping("")}
+}
+
+//BuildQueryData here return a dataAsParam that can be used by to evaluate the variables of the expression
+func (x *ExpressionEngine) BuildQueryData(ctx context.Context, state request.Request) (interface{}, error) {
+	return &dataAsParam{ctx, &RequestDataExtractor{state, x.dataFromReq}}, nil
+}
+
+//BuildReplyData here return a dataAsParam that can be used by to evaluate the variables of the expression
+func (x *ExpressionEngine) BuildReplyData(ctx context.Context, state request.Request, query interface{}) (interface{}, error) {
+	return &dataAsParam{ctx, &RequestDataExtractor{state, x.dataFromReq}}, nil
+}
+
+//BuildRule create a rule for Expression Engine:
+// - first param is one of the action to return
+// - second and following param is a sentence the represent an Expression
+func (x *ExpressionEngine) BuildRule(args []string) (Rule, error) {
+	keyword := args[0]
+	exp := args[1:]
+	e, err := expr.NewEvaluableExpression(strings.Join(exp, " "))
+	if err != nil {
+		return nil, fmt.Errorf("cannot create a valid expression : %s", err)
+	}
+
+	var kind = byte(TypeNone)
+	for k, n := range NameTypes {
+		if keyword == n {
+			kind = byte(k)
+		}
+	}
+	if kind == TypeNone {
+		return nil, fmt.Errorf("invalid keyword %s for a policy rule", keyword)
+	}
+	return &policyRuleExpression{kind, x.kindIfErrorEvaluation, e}, nil
+}
+
+func toBoolean(v interface{}) (bool, error) {
+	if s, ok := v.(string); ok {
+		return strings.ToLower(s) == "true", nil
```

Comment:
why does this to a ToLower()?

Reply:
accept all forms of "TRUE", "True", true", etc ...